File: C/SortingAlgorithm/01-BubbleSort.c
#include <stdio.h>

/*
冒泡排序
冒泡排序（英语：Bubble Sort）是一种简单的排序算法。
它重复地走访过要排序的数列，一次比较两个元素，如果他们的顺序（如从大到小、首字母从A到Z）错误就把他们交换过来。
    这里使用第1个元素与第2个元素比较，使用第2个元素与第3个元素比较...直到倒第2个元素与倒第1个元素比较，将最大的元素拿到最后；
    再使用第1个元素与第2个元素比较，使用第2个元素与第3个元素比较...直到倒第3个元素与倒第2个元素比较，将最大的元素拿到最后倒数第2个位置；
    再使用第1个元素与第2个元素比较，使用第2个元素与第3个元素比较...直到倒第4个元素与倒第3个元素比较，将最大的元素拿到最后倒数第3个位置；
    ...
    传统意义上的冒泡排序；
*/

int main()
{
    int arr[] = {22, 34, 3, 32, 82, 55, 89, 50, 37, 5, 64, 35, 9, 70};

    int len = (int)sizeof(arr) / sizeof(*arr);
    void bubble_sort(int arr[], int len);
    bubble_sort(arr, len);

    int i;
    for (i = 0; i < len; i++)
    {
        printf("%d ", arr[i]);
    }

    printf("\n");

    return 0;
}

void bubble_sort(int arr[], int len)
{
    void swap(int *a, int *b);

    int i, j;
    for (i = 0; i < len - 1; i++)
    {
        for (j = 0; j < len - 1 - i; j++)
        {
            if (arr[j] > arr[j + 1])
            {
                swap(&arr[j], &arr[j + 1]);
            }
        }
    }
}

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

File: Algorithm/Book/漫画算法 小灰的算法之旅/my_implement/第4章 排序算法/01-BubbleSort.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace TestCSharp
{
    class Program
    {
        public static void Main(string[] args)
        {
            Random random = new Random();
            int[] arr = new int[random.Next(29, 43)];
            Parallel.For(0, arr.Length, i => arr[i] = random.Next(0, 100));

            for (int i = 0; i < arr.Length; i++)
                Console.Write($"{arr[i]}, ");

            BubbleSort4(arr);

            Console.WriteLine();
            for (int i = 0; i < arr.Length; i++)
                Console.Write($"{arr[i]}, ");
        }

        /// <summary>
        /// 冒泡排序
        /// </summary>
        /// <param name="arr"></param>
        /// <returns></returns>
        private static void BubbleSort(int[] arr)
        {
            for (int i = 0; i < arr.Length - 1; i++)
            {
                for (int j = 0; j < arr.Length - 1 - i; j++)
                {
                    if (arr[j] > arr[j + 1])
                        swap(arr, j, j + 1);
                }
            }
        }

        /// <summary>
        /// 冒泡排序2
        /// 优化：如果某一轮没有发生任何比较，则取消后面所有轮次的比较
        /// </summary>
        /// <param name="arr"></param>
        private static void BubbleSort2(int[] arr)
        {
            bool flag;  // 如果上一轮没有发生任何比较，flag为true，表示不需要下一轮比较
            for (int i = 0; i < arr.Length - 1; i++)
            {
                flag = true;
                for (int j = 0; j < arr.Length - 1 - i; j++)
                {
                    if (arr[j] > arr[j + 1])
                    {
                        swap(arr, j, j + 1);
                        flag = false;
                    }
                }

                if (flag) break;
            }
        }

        /// <summary>
        /// 冒泡排序3
        /// 优化：如果在border往后的元素，没有发生过交换，则border往后的元素是排序好的，
        ///       则下一轮只要比较到border即可（border可能比 list.Length-1-i 要小）
        /// </summary>
        /// <param name="arr"></param>
        private static void BubbleSort3(int[] arr)
        {
            bool flag;  // 如果上一轮没有发生任何比较，flag为true，表示不需要下一轮比较
            int border = arr.Length - 1;
            for (int i = 0; i < arr.Length - 1; i++)
            {
                flag = true;
                int tempborder = 0;
                for (int j = 0; j < border; j++)
                {
                    if (arr[j] > arr[j + 1])
                    {
                        swap(arr, j, j + 1);
                        tempborder = j;
                        flag = false;
                    }
                }
                border = tempborder;

                if (flag) break;
            }
        }

        /// <summary>
        /// 冒泡排序4
        /// 优化：与BubbleSort3逻辑一致，取消flag（记录上一轮是否发生交换），
        ///       当已排序的边界到达数组的起始位置，就意味着可以取消以后所有轮次的比较
        /// </summary>
        /// <param name="arr"></param>
        private static void BubbleSort4(int[] arr)
        {
            int border = arr.Length - 1;
            for (int i = 0; i < arr.Length - 1; i++)
            {
                int tempborder = 0;
                for (int j = 0; j < border; j++)
                {
                    if (arr[j] > arr[j + 1])
                    {
                        swap(arr, j, j + 1);
                        tempborder = j;
                    }
                }
                border = tempborder;

                if (border <= 0) break;
            }
        }

        private static void swap(int[] arr, int i, int j)
        {
            if (i != j)
            {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
    }
}
